pkg/core/template/engine_test.go

```diff
@@ -191,3 +191,174 @@ func TestEngine(t *testing.T) {
+	return s, nil
+}
+
+func TestResolveTranslations(t *testing.T) {
+	const typeA config.TemplateItemType = "typeA"
+	specA := Spec{
+		Type: typeA,
+		Default: `
+		{
+			"key1": "Hello",
+			"key2": "World"
+		}
+		`,
+	}
+
+	Convey("resolveTranslations", t, func() {
+		test := func(items []config.TemplateItem, expected map[string]map[string]string) {
+			e := NewEngine(NewEngineOptions{
+				TemplateItems: items,
+			})
+			e.Register(specA)
+			e.loader = &mockLoader{}
+
+			actual, err := e.resolveTranslations(typeA)
+			So(err, ShouldBeNil)
+			So(actual, ShouldResemble, expected)
+		}
+
+		// No provided translations
+		test([]config.TemplateItem{}, map[string]map[string]string{
+			"key1": map[string]string{
+				"": "Hello",
+			},
+			"key2": map[string]string{
+				"": "World",
+			},
+		})
+
+		test([]config.TemplateItem{
+			config.TemplateItem{
+				Type:        typeA,
+				LanguageTag: "zh",
+				URI: `
+				{
+					"key1": "你好",
+					"key2": "世界"
+				}
+				`,
+			},
+			config.TemplateItem{
+				Type:        typeA,
+				LanguageTag: "ja",
+				URI: `
+				{
+					"key1": "こんにちは",
+					"key2": "世界"
+				}
+				`,
+			},
+			config.TemplateItem{
+				Type:        typeA,
+				LanguageTag: "en",
+				URI: `
+				{
+					"key1": "Hey"
+				}
+				`,
+			},
+		}, map[string]map[string]string{
+			"key1": map[string]string{
+				"":   "Hello",
+				"en": "Hey",
+				"zh": "你好",
+				"ja": "こんにちは",
+			},
+			"key2": map[string]string{
+				"":   "World",
+				"zh": "世界",
+				"ja": "世界",
+			},
+		})
+	})
+}
+
+func TestMakeLocalize(t *testing.T) {
+	key := "key1"
+	Convey("makeLocalize", t, func() {
+		test := func(m map[string]string, preferredLanguageTags []string, expected string) {
+			localize := makeLocalize(preferredLanguageTags, map[string]map[string]string{
+				key: m,
+			})
+			actual, err := localize(key)
+			So(err, ShouldBeNil)
+			So(actual, ShouldEqual, expected)
+		}
+
+		// Select default if there is no preferred languages
+		test(map[string]string{
+			"":   "Hello from default",
+			"en": "Hello from en",
+			"ja": "Hello from ja",
+			"zh": "Hello from zh",
+		}, nil, "Hello from default")
+
+		// Select default if there is no preferred languages
+		test(map[string]string{
+			"":   "Hello from default",
+			"en": "Hello from en",
+			"ja": "Hello from ja",
+			"zh": "Hello from zh",
+		}, []string{}, "Hello from default")
+
+		// Simply select japanese
+		test(map[string]string{
+			"":   "Hello from default",
+			"en": "Hello from en",
+			"ja": "Hello from ja",
+			"zh": "Hello from zh",
+		}, []string{"ja-JP", "en-US", "zh-Hant-HK"}, "Hello from ja")
+
+		// Select the default because japanese and chinese is unknown to a korean speaker.
+		test(map[string]string{
+			"":   "Hello from default",
+			"ja": "Hello from ja",
+			"zh": "Hello from zh",
+		}, []string{"kr-KR"}, "Hello from default")
+
+		// Select english because english is an international language.
```

Comment:
Can you explain more on these three test cases? The term 'international language' and 'best variant' is not familiar to me, and looks rather 'magical'.

Reply:
I don't know either. The comment given there is my guess. [This blog](https://blog.golang.org/matchlang) also just gives some examples. The blogpost says that [the implementation](https://github.com/golang/text/blob/master/language/match.go) takes CLDR data into account. So it is magical because it is driven by real world data.

For example, I guess there are more Taiwanese than Hongkongers. So `zh-Hant-TW` is preferred over `zh-Hant-HK`

Reply:
I suppose so... I just seems unintuitive that a missing language would resolve to English localization, rather than default localization.

I think we can omit the three test cases, since they're actually testing the underlying CLDR data, and I'm not sure if language/system upgrade would change the behavior. We can just test the most expected cases (e.g. `ja-JP` falls back to `ja`, `zh-Hant` would not falls back to `ja`, etc.).

Reply:
Agree. Let me remove the test case related to the behavior of Matcher.